src/main/java/com/marklogic/kafka/connect/source/DslQueryHandler.java

```diff
@@ -43,10 +47,13 @@ protected String injectConstraintIntoQuery(String previousMaxConstraintColumnVal
         if (previousMaxConstraintColumnValue != null) {
             String constraintPhrase = ".where(op.gt(op.col(\"" + constraintColumnName + "\"), \"" + previousMaxConstraintColumnValue + "\"))";
             String originalDslNoWhitespace = QueryHandlerUtil.removeWhitespace(userDslQuery);
-            if (originalDslNoWhitespace.contains(").")) {
-                int firstClosingParen = originalDslNoWhitespace.indexOf(").");
-                constrainedDsl = originalDslNoWhitespace.substring(0, firstClosingParen + 1)
-                    + constraintPhrase + originalDslNoWhitespace.substring(firstClosingParen + 1);
+
+            Pattern pattern = Pattern.compile(BALANCED_PARENS_REGEX_PATTERN);
```

Comment:
This logic is a little complex (in the sense of - I don't think it's immediately clear what's going on here), and I think a separate method would be nice, both for explaining what's going on and in case you want to unit test it directly - e.g.

    protected int findEndPositionOfFirstAccessor(originalDslNoWhitespace)

The benefit of that would be very targeted unit-testing of the very-complicated regex from stackoverflow. 

Even if it's not unit-tested directly, it's nice for self-documenting purposes and also simplifies the overall method. 

I was thinking an "Optional" would be nice here too to make it clear that something might not be found:

    protected Optional<Integer> find...

Reply:
Actually, I have removed the if statement, since if the matcher doesn't find anything, that would be a runtime error.
So, I'm not using Optional here.